### DISCUSSION TITLE
Is it possible to have volume and/or pan control for each TG?

### DISCUSSION THREAD

**Discussion by iluvsa** [probonopd/MiniDexed#702] [General]
Hey All,

I've managed to get cc MIDI signals from an Arturia MiniLab to affect the MiniDexed - the pads can be used as buttons and the knobs give me continuous change over Reverb level, Detune, Pan, Volume etc. What would be really amazing is if it was possible to control individual TG Volume and Pan, for instance, with it's own knob on the MiniLab - I mean, there are eight of them on each row so this would be so perfect! I think because they are all on the same MIDI channel (so that sound together), that this is not possible, but I just wanted to make sure some smart folks hadn't figured out how to do this that I was unaware of!!! Thanks!

**Comment by diyelectromusic** [comment]
It's not currently possible with all TGs on the same channel, but that has been asked about before.  We need to think how best to handle "system" level MIDI configurations like this that would end up having to largely outside the individual TG configuration.

Kevin

**Reply by iluvsa** [reply]
Thanks Kevin I wanted to make sure I wasn’t missing something - I am gonna try to get creative and see if I can’t figure out a workaroud….stay tuned!

**Reply by diyelectromusic** [reply]
I did wonder if the SysEx device number might have allowed all TGs to receive a SysEx parameter change message for performance parameter 16 (total volume) but at present we check against that and compare it to the TG MIDI channel (https://github.com/probonopd/MiniDexed/blob/e03122a3d232e7b20e5ccbb54de6f0dff5f7666f/src/mididevice.cpp#L246).

From the MIDI Technical Data Reference Manual:
![image](https://github.com/user-attachments/assets/c2b39071-6eb6-4222-a0ee-0e7d4d4d4e76)
![image](https://github.com/user-attachments/assets/f764af58-5126-437f-ac1d-dbba4c1ce9d4)

We could decide to break that link, then SysEx would be received by the whole system and directed to the device specified in that parameter regardless of MIDI channel of the TG...

But I don't know how a real DX7 handles the device number and the manual only states:
![image](https://github.com/user-attachments/assets/7505a421-66a7-40ba-bc3d-de5cc68a8484)

So I guess we'd need to think about it properly.

Another possibility might be if the original button UI can be triggered via SysEx using the "remote switch" option (gr=6, h=3 I think...):
![image](https://github.com/user-attachments/assets/ae045fae-2d8c-4dba-a743-a2336b75fe6e)

But again, without knowing what combinations of buttons might correspond to volume, that probably still suffers from the device number problem too.

Aside, I think the reception requirements flowchart has a mistake:
![image](https://github.com/user-attachments/assets/9b8a6163-552d-4b4f-bd4d-40c86c8910be)

There are two sets of $F0 $43 $1S $18.  I think the second one is meant to be $1B not $18 if it is to match the remote switch options above...

Kevin

**Reply by diyelectromusic** [reply]
Ok, maybe not.  Looks like we can only accept parameter groups 0 or 2, which correspond to the original DX7 MIDI implementation - this is from The Complete DX7:
![image](https://github.com/user-attachments/assets/ea31a65a-3969-4b89-83b4-184604b4159c)

We don't currently handle any group 6 parameters as in the DX7II which includes the performance parameters.

Oh well.  All stuff that could be added I guess...

Kevin

**Reply by diyelectromusic** [reply]
Looks like I'd already noticed there was something odd about device ID... more here: https://github.com/probonopd/MiniDexed/issues/502

The more I think about this, the more I think it ought to be the TG number and not related to MIDI channels at all...

I wonder how we can find out more about how it was meant to be used?

**Reply by iluvsa** [reply]
This is all really great information, though! I was gonna try and set up sort of a MIDI sequence of events, like press one button, get 3 separate midi events to happen (blokas midihub does this well), that gets us into individual TG volumes, and then have a single knob to turn up or down volume on that TG, but there is no reference to where you are in the Menu options, so it doesn’t always land you at the TG Volume. You’s have to hit the “Home” button each time which is less than ideal…

**Reply by diyelectromusic** [reply]
My SysEx UI control (or the alternative NRPN version here https://github.com/probonopd/MiniDexed/pull/604) is the way to go... I just need to implement it properly :)

But I think adding individual CCs for the volume of each TG is pretty straight forward.  I'm just chewing over what else we could use like that too - you mentioned pan too... anything else do you think?

Vol via the general purpose CC messages is pretty straight forward - that could be a single config option to just turn it on.  It would be recognised on any MIDI channel being listened to by any TG mind, but that's probably not a major issue.

Kevin

**Reply by iluvsa** [reply]
> But I think adding individual CCs for the volume of each TG is pretty straight forward. I'm just chewing over what else we could use like that too - you mentioned pan too... anything else do you think?

I think the Detune for individual TG's would be super cool too, and might be even more useful than pan? Reverb is nice but I don't get the immediate need to set that per TG. Detune could get us some more dystopian CS-80 vibes a la Rachael, which I feel like we will need in the near future....;P

**Comment by diyelectromusic** [comment]
Thinking about this, there are a few ways to implement "one per TG" global MIDI controllers. One option is simply to have 8 config entries in minidexed.ini for each parameter, e.g.
MIDI_CC_VOL_TG1
MIDI_CC_VOL_TG2
...
MIDI_CC_PAN_TG1
...
MIDI_CC_DETUNE_TG1
...
etc.

And do the same for any new options we'd like to support too in the future.

But another option might be to configure several fixed maps of MIDI CC messages that can then be used for any parameters we want to support, and potentially any future ones too.  Then all that is needed is a single config entry per control parameter to specify which "map" we're going to use.

Eg:
MIDISystemCCVol = n
MIDISystemCCPan= n
MIDISystemCCDetune = n

where n = an index representing a specific CC map of 8 CC messages - one per TG.  Some example maps could be:
0 = disabled
1 = CC 16-19, 80-83 - the General Purpose Controllers 1-8
2 = CC 20-27
3 = CC 52-59
4 = CC 102-109
5 = CC 110-117
6 = CC 3, 9, 14-15, 38-31
7 = CC 35, 41, 46-47, 60-63

Options 2-7 are all various sets of undefined CC parameters I believe.

The first gives total flexibility and is probably simpler to code, but there will be lots of repeated code reading a config option per CC (x8 for all TGs, x number of controls).

The second is more limited, with hard-coded options, and is perhaps slightly more complex to code, but I think a lot easier for a user to configure and it will be easier to add new controls in the future once all the code is done.

I'm minded to look into how to do the second at the moment, but does anyone else have a view?

The other thing to chew over of course, is are these "performance" settings or minidexed.ini settings?  Whilst it would be nice to be able to load up different mappings for different performances, in reality our performances are more about the sounds than the MIDI config. I think it would be quite a pain to have these duplicated in performance files.

So if they are minidexed.ini settings, again to avoid cluttering it too much, I think I prefer the idea of a single config option per control rather than 8 - so the second option again seems better to me.

Kevin

**Reply by diyelectromusic** [reply]
There is an initial implementation of the "map" idea here: https://github.com/probonopd/MiniDexed/pull/704 for people to try out.  Let me know what you think and if this has any merit as an idea.

Kevin

**Reply by iluvsa** [reply]
Nice Kevin, I will try this when I get home tonight! I think the way it is implemented is well laid out in terms of theory above, and I like the simplicity of it, particularly because, for instance, most modern usb controllers allow you to assign the CC number, so you can essentially create the controller bespoke for you using one of the "presets" you lay out.

The only downside I can see is, if you were to allow the user to assign their own cc's to each TG, one COULD do some fun things like control two TGs with one cc message, so for instance, if you had a smaller controller with less knobs, or just wanted groups of TG's (pianos, strings) for quicker performative changes, you could do that easier.

That being said, I am so excited to go home and try out the individual TG layout and report back!

**Reply by diyelectromusic** [reply]
Well as I say, we could go for a "detail all the CC settings" approach, but that would be 24 settings to do vol, pan and detune (assuming someone wanted them all!) plus another 8 for any future ideas we have.

The single-channel versions still work too, so there might still be some MIDI mapping possibilities there, but the permutations are already getting a little hard to keep track of :)

Kevin

**Comment by diyelectromusic** [comment]
This is now available and wiki updated with the options.

Kevin
